### ISSUE TITLE
Paging and large collections

### ISSUE THREAD

**Issue by elfatherbrown** [directus/gatsby-source#35]
Hey guys. Great work with all this.
I have a use case that involves thousands of records whilst directus will return a max of 300 for any get_items call by default. Limitless searches require a -1 in the limit query field, but I do understand this might not be the best strategy for the plugin.

For now im just using a modified version as I have to get this done asap.I include this change at the end for reference. 

However, i guess the most sensible thing would be to allow for a config entry, maybe per collection, to set the limit of records fetched (unless there is something in building the content mesh i still dont understand, which would be very very likely as im a total newb and I dont tscript at all).

If or when I have the time, ill try to make this work the right way and submit a patch. 

PS: Sorry if my ranting is just taking valuable time away from you... :/

```
public async getCollectionRecords(collection: string): Promise<any[]> {
    try {
      log.info(`Fetching records for ${collection}...`);

      const { data: items = [] } = (await this._api.getItems(collection, {
        fields: '*.*',
        limit:-1, //<------RIGHT HERE
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      })) as { data: any[] };

      return items.filter(record => this._shouldIncludeRecord(record, collection));
    } catch (e) {
      log.error(`Failed to fetch records for collection "${collection}"`);
      log.error(`Did you grant READ permissions?`);
      throw e;
    }
  }
```

**Comment by JulioC** [comment]
I've just spent a few hours trying to debug a related issue:

When loading collections with images as relations, some of the items were missing their images in the GraphQL data. After a long debug session, I've found out the plugin was skipping `addRelation` (in `ContentRelation._updateTable`) because the related file was not found in `directus_files` collection.

It turns out only the first 200 files were being loaded by `DirectusService.getAllFiles` (due to the default limit set in Directus global settings).

To avoid leaving this pitfall in the plugin, I'd suggest adding a `limit` option, with default value of -1. This limit would be applied to all requests in DirectusService

**Comment by loganwishartcraig** [comment]
Hey, this is great, thank you for reporting!

We'll definitely need to add something to resolve this. I like the idea of setting a ```limit``` option with per collection overrides available.

@jooola What do you think a sensible default for the limit is? Should we go for ```"-1"``` to include all records, or ```undefined``` to use the Directus default? I'm leaning towards ```"-1"``` myself, where they could set to ```undefined``` manually to return to the default Directus behavior.

**Comment by jooola** [comment]
@loganwishartcraig We could assume people working with this gatsby-source package are the also the Directus server admins. So setting the limit to `-1` is ok for me as they know what they are doing, still I feel like this isn't clean and more like a workaround.

We should consider 2 things: Rate limiting and Query timeout. Those settings a linked to server and client performances.
We cannot run 100 query/second which might trigger some rate limiting somewhere, but we shouldn't run a single request to query 100k heavy items.

What could be implemented is to add a limit setting, with default to undefined (delegate to the SDK), add to keep querying if some data hasn't been fetched yet (pagination).

So as quick fix we can set the limit to -1, but in the long run I would see something more dynamic (maybe the sdk could implement this, not us).

**Comment by loganwishartcraig** [comment]
@jooola Just for clarity, you're thinking is we'd expose two config options? A ```rateLimit``` defining the maximum number of records returned for each query, and a ```queryThrottle``` defining the how long we'd wait between API requests?

Would be cool if pagination were naively supported by the SDK, but I don't think it would be too hard to implement on our end here in the plugin. The API seems like it currently returns all the necessary information. As long as we pull all the records into memory eventually - before trying to build the Gatsby nodes - it should be fine.

**Comment by jooola** [comment]
Wasn't thinking about query throttle, but why not. That's a good idea. I was more thinking about HTTP server timeout.

That said I don't know which option is better, query a lot, or a lot of queries. Maybe I'm looking for an false problem.

**Comment by JulioC** [comment]
The accepted PR doesn't change anything related to `getAllFiles` method, so the bug will still happens if you have too many media files in Directus

**Comment by loganwishartcraig** [comment]
@JulioC The other calls are addressed in a subsequent commit f43fc6e which was included in the release

**Comment by loganwishartcraig** [comment]
I'll make a note about that in the PR chain
